actionview/lib/action_view/helpers/tag_helper.rb

```diff
@@ -46,8 +49,168 @@ class TagBuilder # :nodoc:
         include CaptureHelper
         include OutputSafetyHelper
 
-        HTML_VOID_ELEMENTS = %i(area base br col embed hr img input keygen link meta param source track wbr).to_set
-        SVG_SELF_CLOSING_ELEMENTS = %i(animate animateMotion animateTransform circle ellipse line path polygon polyline rect set stop use view).to_set
+        def self.define_element(name, code_generator:, method_name: name.to_s.underscore)
+          code_generator.define_cached_method(method_name, namespace: :tag_builder) do |batch|
+            batch.push(<<~RUBY) unless instance_methods.include?(method_name.to_sym)
+              def #{method_name}(*arguments, **options, &block)
+                tag_string(#{name.inspect}, *arguments, **options, &block)
+              end
+            RUBY
+          end
+        end
+
+        def self.define_void_element(name, code_generator:, method_name: name.to_s.underscore, self_closing: false)
+          code_generator.define_cached_method(method_name, namespace: :tag_builder) do |batch|
+            batch.push(<<~RUBY)
+              def #{method_name}(content = nil, escape: true, **options, &block)
+                if content.nil? && block.nil?
+                  void_tag_string(#{name.inspect}, options, escape, #{self_closing})
+                else
+                  tag_string(#{name.inspect}, content, escape: escape, **options, &block)
+                end
+              end
+            RUBY
+          end
+        end
+
+        def self.define_self_closing_element(name, **options)
+          define_void_element(name, self_closing: true, **options)
+        end
+
+        ActiveSupport::CodeGenerator.batch(self, __FILE__, __LINE__) do |code_generator|
+          with_options(code_generator: code_generator) do
+            define_void_element :area
```

Comment:
@casperisfine I wasn't familiar with CodeGenerator. Thank you for teaching me about that one!

I've wrapped the definitions in a single batch, and have also added a similar block to the `method_missing` optimization. Is that what you had in mind?